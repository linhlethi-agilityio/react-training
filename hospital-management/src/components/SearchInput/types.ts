import { IInputProps } from '@components/Input/types';
import { ChangeEvent } from 'react';

export interface ISearchInputProps {
  className?: string;
  searchText: string;
  location: string;
  locationInputConfig: IInputProps;
  searchInputConfig: IInputProps;
  onChangeLocation?: (e: ChangeEvent<HTMLInputElement>) => void;
  onChangeSearch?: (e: ChangeEvent<HTMLInputElement>) => void;
  onSearch?: () => void;
}
