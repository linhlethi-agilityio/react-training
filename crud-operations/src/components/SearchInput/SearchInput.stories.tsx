import type { Meta, StoryObj } from '@storybook/react';

// Components
import SearchInput from '.';

const meta: Meta<typeof SearchInput> = {
  component: SearchInput,
};

export default meta;
type Story = StoryObj<typeof SearchInput>;

export const Default: Story = {
  args: {
    placeholder: 'Start typing...',
    onChange: () => null,
    value: 'Student',
  },
};
