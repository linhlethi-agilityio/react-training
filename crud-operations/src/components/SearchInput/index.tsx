import { memo } from 'react';
import { Input, InputProps, InputGroup, InputRightElement } from '@chakra-ui/react';

// Components
import { SearchIcon } from '@icons';

interface SearchInputProps extends InputProps {
  placeholder?: string;
  onChange: (e: React.ChangeEvent<HTMLInputElement>) => void;
  value: string;
}

const SearchInput = ({ placeholder = 'Start typing...', onChange, value, ...rest }: SearchInputProps) => {
  return (
    <InputGroup maxW={212}>
      <Input
        borderRadius="sm"
        focusBorderColor="transparent"
        _placeholder={{
          color: 'text.placeholder',
        }}
        placeholder={placeholder}
        onChange={onChange}
        value={value}
        {...rest}
      />
      <InputRightElement style={{ cursor: 'pointer' }}>
        <SearchIcon />
      </InputRightElement>
    </InputGroup>
  );
};

export default memo(SearchInput);
