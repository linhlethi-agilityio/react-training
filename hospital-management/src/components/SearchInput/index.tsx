import styled from 'styled-components';
import { ISearchInputProps } from './types';
import { LocationIcon, SearchIcon } from '@components/Icons';
import { StyledInput } from '@components/Input';
import { FormEvent, memo, useCallback } from 'react';

const SearchInput = memo(
  function SearchInput(props: ISearchInputProps): React.ReactElement {
    const {
      className,
      onChangeLocation,
      onChangeSearch,
      onSearch,
      searchText,
      location,
      locationInputConfig,
      searchInputConfig,
    } = props;

    const handleSubmit = useCallback(
      (e: FormEvent<HTMLFormElement>) => {
        e.preventDefault();
        onSearch?.();
      },
      [onSearch],
    );

    return (
      <form className={className} onSubmit={handleSubmit}>
        <StyledLocation>
          <LocationIcon size={11} />
          <StyledInput
            {...locationInputConfig}
            variant="primary"
            onChange={onChangeLocation}
            value={location}
          />
        </StyledLocation>
        <StyledSearch>
          <button type="submit">
            <SearchIcon size={16} />
          </button>
          <StyledInput
            {...searchInputConfig}
            variant="primary"
            onChange={onChangeSearch}
            value={searchText}
          />
        </StyledSearch>
      </form>
    );
  },
  (prev, next) => prev.location === next.location && prev.searchText === next.searchText,
);

const StyledSearchInput = styled(SearchInput)<ISearchInputProps>`
  display: flex;
  background-color: #fcfcfc;

  button {
    border: none;
    background: none;
    cursor: pointer;
  }
`;

const StyledLocation = styled.div`
  border: 1px solid #4c84c3;
  display: flex;
  border-right: none;
  border-radius: 10px 0px 0px 10px;
  padding: 10px 39px 10px 11px;
`;

const StyledSearch = styled.div`
  border: 1px solid #4c84c3;
  display: flex;
  border-radius: 0px 10px 10px 0px;
  padding: 10px 122px 10px 17px;
`;

export { StyledSearchInput };
