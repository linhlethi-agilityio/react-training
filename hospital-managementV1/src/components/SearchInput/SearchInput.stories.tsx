import { ComponentMeta, ComponentStory } from '@storybook/react';
import { StyledSearchInput } from '@components/SearchInput';
import { ChangeEvent, useState } from 'react';
import { action } from '@storybook/addon-actions';

export default {
  title: 'SearchInput',
  component: StyledSearchInput,
  decorators: [
    (Story, context) => {
      const { args } = context;

      const [searchValue, setSearchValue] = useState({ location: '', search: '' });

      const handleChange = (e: ChangeEvent<HTMLInputElement>) => {
        const { name, value } = e.target;

        setSearchValue({
          ...searchValue,
          [name]: value,
        });
      };

      return (
        <Story
          {...args}
          searchText={searchValue.search}
          location={searchValue.location}
          onChange={handleChange}
        />
      );
    },
  ],
} as ComponentMeta<typeof StyledSearchInput>;

const Template: ComponentStory<typeof StyledSearchInput> = (args) => {
  return <StyledSearchInput {...args} />;
};

export const SearchInput = Template.bind({});

/**
 * searchText and location in decorators
 */
const searchInputProps = {
  locationInputConfig: {
    name: 'location',
    placeholder: 'Location',
    inputWidth: '80px',
  },
  searchInputConfig: {
    name: 'search',
    placeholder: 'Search for Doctors, hospital & others....',
    inputWidth: '261px',
  },
  onSearch: () => action('onSearch')(),
};
SearchInput.args = searchInputProps;
