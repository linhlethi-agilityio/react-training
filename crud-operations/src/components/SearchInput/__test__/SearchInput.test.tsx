import { fireEvent, render } from '@test-utils';
import '@testing-library/jest-dom';

// Components
import SearchInput from '..';

const mockOnChange = jest.fn();

const mockProps = {
  onChange: mockOnChange,
  value: '',
};

describe('SearchInput component', () => {
  test('should match snapshot for SearchInput', () => {
    const { container } = render(<SearchInput {...mockProps} />);

    expect(container).toMatchSnapshot();
  });

  test('call onChange func when change input', () => {
    const { getByPlaceholderText } = render(<SearchInput placeholder="search..." {...mockProps} />);

    const inputElement = getByPlaceholderText('search...');

    fireEvent.change(inputElement, { target: { value: 'Test input' } });

    expect(mockOnChange).toHaveBeenCalledTimes(1);
    expect(mockOnChange).toHaveBeenCalledWith(expect.any(Object));
  });
});
