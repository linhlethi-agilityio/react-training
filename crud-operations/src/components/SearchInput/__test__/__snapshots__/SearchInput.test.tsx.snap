// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchInput component should match snapshot for SearchInput 1`] = `
<div>
  <div
    class="chakra-input__group css-10ip56e"
    data-group="true"
  >
    <input
      class="chakra-input css-2pjcr"
      placeholder="Start typing..."
      value=""
    />
    <div
      class="chakra-input__right-element css-dp9r5f"
      style="cursor: pointer;"
    >
      <svg
        height="14"
        viewBox="0 0 14 14"
        width="14"
      >
        <path
          d="M13.904 13.169 10.377 9.64a.325.325 0 0 0-.232-.095h-.282A5.687 5.687 0 0 0 5.688 0 5.687 5.687 0 0 0 0 5.687a5.687 5.687 0 0 0 9.546 4.179v.279c0 .087.035.17.095.232l3.528 3.527a.329.329 0 0 0 .465 0l.27-.27a.329.329 0 0 0 0-.465ZM5.688 10.5A4.81 4.81 0 0 1 .875 5.687 4.81 4.81 0 0 1 5.688.875 4.81 4.81 0 0 1 10.5 5.688 4.81 4.81 0 0 1 5.687 10.5Z"
          fill="#C4C4C4"
        />
      </svg>
    </div>
  </div>
</div>
`;
